library/core/src/iter/adapters/step_by.rs

```diff
@@ -233,3 +225,96 @@ where
 // StepBy can only make the iterator shorter, so the len will still fit.
 #[stable(feature = "iterator_step_by", since = "1.28.0")]
 impl<I> ExactSizeIterator for StepBy<I> where I: ExactSizeIterator {}
+
+trait SpecRange: Iterator {
+    fn spec_fold<Acc, F>(self, acc: Acc, f: F) -> Acc
+    where
+        F: FnMut(Acc, Self::Item) -> Acc;
+}
+
+impl<I> SpecRange for StepBy<I>
+where
+    I: Iterator,
+{
+    default fn spec_fold<Acc, F>(mut self, mut acc: Acc, mut f: F) -> Acc
+    where
+        F: FnMut(Acc, Self::Item) -> Acc,
+    {
+        #[inline]
+        fn nth<I: Iterator>(iter: &mut I, step: usize) -> impl FnMut() -> Option<I::Item> + '_ {
+            move || iter.nth(step)
+        }
+
+        if self.first_take {
+            self.first_take = false;
+            match self.iter.next() {
+                None => return acc,
+                Some(x) => acc = f(acc, x),
+            }
+        }
+        from_fn(nth(&mut self.iter, self.step)).fold(acc, f)
+    }
+}
+
+macro_rules! spec_int_ranges {
+    ($($t:ty)*) => ($(
+        impl SpecRange for StepBy<Range<$t>> {
+            #[inline]
+            fn spec_fold<Acc, F>(self, mut acc: Acc, mut f: F) -> Acc
+            where
+                F: FnMut(Acc, Self::Item) -> Acc
+            {
+                use crate::convert::TryFrom;
+
+                let mut range = self.iter;
+
+                if !self.first_take {
+                    // `next()` implements `increment-by-step, get-and-increment-by-one`, leaving
+                    // range.start set to one element after the previously returned one.
```

Comment:
It doesn't have to -- https://github.com/rust-lang/rust/pull/52193/files FCP'd that it's allowed to have `start` be the next thing to return after a `next()`.

Would it be worth using that freedom here?

Reply:
That would mean changing or specializing all forward iteration methods to have that behavior, including `size_hint`.

With `advance_by` we can now do that in a reasonable manner but that's a bigger change and one that impacts observable behavior (even if allowed) so that should be done separately imo.

Reply:
Ok, turns out this was necessary anyway. I have no specialized all methods so we can operate differently on `Range<{integer}>`.